ice/settings.py

```diff
@@ -27,10 +32,45 @@ class Settings(BaseSettings):
     OUGHT_ICE_AUTO_BROWSER: bool = True
     PAPER_DIR: Path = Path(__file__).parent.parent / "papers"
 
+    # note these attributes are read differently- see [__getattribute__]
+    OPENAI_API_KEY: str = ""
+    OUGHT_INFERENCE_API_KEY: str = ""
+    ELICIT_AUTH_TOKEN: str = ""
+
+    def __get_and_store(self, setting_name: str, prompt: Optional[str] = None) -> str:
+        # We use [__getattribute__] to read these attributes, so that we can
+        # prompt the user for them if they are not already set.
+        if prompt is None:
+            prompt = f"Enter {setting_name}: "
+        if self.__dict__[setting_name] == "":
+            # TODO someday: one way this stinks is that starting the server will
+            # bury this prompt. (the server is started in a background process
+            # and has some log messages). i don't know if there's a clean way
+            # to fix this without re-architecting some stuff.
```

Comment:
Good point. How about creating a global `log_lock = threading.Lock()` and wrapping both the line below and the body of `_server_and_browser` in `with log_lock:`?

Reply:
👍 d011138